### PR TITLE
Batching buffer with minChunkSize

### LINKED ISSUE #32546: ResteasyReactiveOutputStream could make better use of the Netty (direct) pooled allocator
### Description

`quarkus.resteasy-reactive.output-buffer-size` (see [quarkus.resteasy-reactive.output-buffer-size](https://quarkus.io/guides/all-config#quarkus-resteasy-reactive-common_quarkus.resteasy-reactive.output-buffer-size)'s doc) seems to not just control when responses should contain the `Content-Length` header and being chunked, but the initial eagerly allocated capacity of the buffer used to stream content over the connection per response as well, see https://github.com/quarkusio/quarkus/blob/f6851e3371bf3b28c394a186964b1625f8be5ce2/independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/ResteasyReactiveOutputStream.java#L208

This "eager" behavior, despite seems optimal while simple (because it allows to perform optimistic batching) can be problematic for the Netty allocator, due to how it works under the hood. 
Adding more notes below, to explain the effects/consequences.

**NOTE**:
the default buffer size is `8191` bytes 

eg

Performing an allocation of 8K and assuming a I/O caller thread (aka the event loop), in a default configured Netty scenario, is going to use the netty `PoolThreadCache` at https://github.com/netty/netty/blob/c353f4fea52559d09b3811492c92a38aa1887501/buffer/src/main/java/io/netty/buffer/PoolThreadCache.java#L299-L303 ie a so-called `small` direct pooled allocation.

The cache itself is organized in order to have few `MemoryRegionCache`s (in an array) chosen based on the normalized required capacity; in this case it's going to use the one at the `sizeIdx = 31`, obtained by (https://github.com/netty/netty/blob/eb3feb479826949090a9a55c782722ece9b42e50/buffer/src/main/java/io/netty/buffer/SizeClasses.java#L317)[SizeClasses::size2SizeIdx].

Every `MemoryRegionCache` has a finite number of pooled (and thread local) buffers ie `io.netty.allocator.smallCacheSize` , which is 256 by default.

Such pooled and thread local instances are not filled into `MemoryRegionCache` till they are used/allocated the very first time (on demand) - it means that in the worst case; given that we *always* allocate 8K chunks, we risk to have a (8K * 256) memory footprint per even loop thread, while idle, if a previous run has caused them to be fully allocated (maybe because of 256 concurrent and slow connections running on the same event loop thread).

In short: the problem is not the allocation cost, because, in the happy path is a cached already-happened thread-local allocation, but is a matter of idle memory footprint and bad utilization of the existing caching behavior provided by Netty: if we could use different sized allocations here, we can make uses of all the different `sizeIdx` that Netty provide, possibly reducing the idle utilization as well due to retaining just the effectively used ones (or near to what has been the effective usage).
In addition, using many `MemoryRegionCache` entries, reduce the chances of un-happy paths, because we are not bounded just to a single `MemoryRegionCache` capacity (that's fairly low, as said - just 256 buffers).

### Implementation ideas

https://github.com/franz1981/quarkus/tree/append_buffer

### PR DESCRIPTION
**Pull request by franz1981**
Fixes #32546 

This is still in draft due to few still opened questions:
1. what happen to the batch buffer content in case of partial failures within `flushBuffer`? it should cleanup its whole content, loosing it? What's supposed to do the caller?
2. what happen to `ResteasyReactiveOutputStream` in case of failure while using the buffer returned by `flushBuffer`? releasing it if not already released? (in the previous impl it wasn't well defined in each state)
3. why 128? We would like to expose even that configuration parameters to users? It's great for benchmarking tests, but maybe too complex for avg urs @geoand ?
4. how performance looks like? Is any better? [WIP - after or during Quarkus F2F]

### CODE REVIEW

**Comment by franz1981** [comment]
@Sanne @geoand @stuartwdouglas the implementation can already be reviewed actually - apart from the questions at https://github.com/quarkusio/quarkus/pull/32858#issue-1681170590

**Comment by franz1981** [comment]
If this would work @Sanne the change at https://github.com/TechEmpower/FrameworkBenchmarks/pull/7861/files#diff-ea8995e85f3d4bb82b2c8aa7feab84fb8d193519e3f919c7d43c3c2c25310344R31-R32 won't be impactfull anymore and real users as well would benefit from not using 8K Netty (pooled) direct buffer sized for each response, as now

**Comment by geoand** [comment]
Thanks @franz1981! 

Let's move this ahead in a couple weeks time :)(

File: independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/AppendBuffer.java
Comment (franz1981) [inline independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/AppendBuffer.java:133]: Probably worth dropped the whole existing content (into the queue) here: it's an all-or-nothing use case

**Comment by geoand** [comment]
I'd like to see how this performs with small / medium and large payloads compared to what we have currently

**Comment by geoand** [comment]
Many small ones would only happen in SSE or other streaming scenarios, definitely not a major concern

**Comment by franz1981** [comment]
I've spent few weeks testing with our CI using Techempower for evident regressions with both stacks (reactive/blocking) with/without hibernate and I didn't had any performance regression nor improvement, throughput-wise.

What this patch can deliver instead, is a very tangible improvement in the off-heap/direct memory usage in case of blocking endpoints, instead:
- JSON encoding happen from within the EQE threads (aka the blocking thread pool), causing the allocation of direct `ByteBuf` to happen there
- EQE threads while writing to the wire, it delegates an async task on the Vertx/Netty event loop (which surprise me a bit, given that it could use a `write/writeAndFlush` instead of a generic `Runnable` on `ConnectionBase::queueForWrite`), see below:

image

- that means that the buffer allocated in EQE can be alive for some time till flushed to the I/O and making the EQE ready to pick other requests in-flight

This last behavior make very important to NOT over-size such allocation buffer, because it determines how much overall in-flight bytes can be required to be allocated  to the Netty allocator, that will grow to accomodate the incoming load.
This same behavior won't happen for buffers allocated within the I/O threads, instead which interact directly with I/O and can be reused right after for the same or another connection.

In a JSON hello world load test using a blocking endpoint the saving of RSS has been of ~2MB vs ~70MB.

**Comment by geoand** [comment]
> In a JSON hello world load test using a blocking endpoint the saving of RSS has been of ~2MB vs ~70MB.

This is a very significant improvement! Definitely makes the PR worthy of inclusion :).

Thanks a lot of this!

**Comment by franz1981** [comment]
I've just to decide what's the `minChunkSize` by default: Netty allocations capacity are
```
16,32,48,64,80,96,112,128,...
```
Meaning that in theory it should be 16, to be friendly with Netty

**Comment by Sanne** [comment]
What is holding this up? Any concerns?

**Comment by geoand** [comment]
Yup, I've already talked with @franz1981 and once he is ready, we can rebase onto `main`, take it out of draft and merge when CI goees greem

File: independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/AppendBuffer.java
Comment (franz1981) [inline independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/AppendBuffer.java:114]: I didn't rollback the lastBuffer's writer Index either here (in case it has been written) :(

**Comment by franz1981** [comment]
The original code, given that was using a single buffer, was correctly handling releases of resources, so it has taken me a while to ensure a similar behavior now that we can have few chunks instead....
I didn't add any unit test on `AppendBuffer` unless we plan to use it elsewhere.

@geoand 
Some thoughts that I would like to solve before merging this:
- as said, any append happening from outside the I/O threads, can lead to unbounded grow of direct memory (impacting RSS): right now I'm capping the minimum allocation to be 128 bytes (by default), but other strategies could be to always be `16` bytes or, more adaptively, `128` bytes if the `append` happen in an I/O thread (because from I/O threads, I/O is unlikely to accumulate, unless network congestion), or `16` if is not on an I/O thread (to avoid uncontrolled memory growth).

**Review by stuartwdouglas — APPROVED** [review]

**Comment by geoand** [comment]
I don't think we need this to be configurable - at least not for the time being.
It's a very low level mechanism and I don't really see us needing to tune it.

If we do in the future come up with specific scenarios (I mean code that users actually run) were configuration would be beneficial, we can add it (along with some good documentation).

**Review by geoand — APPROVED** [review]

**Comment by geoand** [comment]
Force pushed an update that makes the `minChunkSize` configurable via: `quarkus.resteasy-reactive.min-chunk-size`

**Comment by geoand** [comment]
Merging, thanks a ton for this @franz1981!

**Pull request merged**
